src/base_alloc/base_alloc_linear.c

```diff
@@ -69,26 +105,59 @@ umf_ba_linear_pool_t *umf_ba_linear_create(size_t pool_size) {
 
 void *umf_ba_linear_alloc(umf_ba_linear_pool_t *pool, size_t size) {
     size_t aligned_size = align_size(size, MEMORY_ALIGNMENT);
-
     util_mutex_lock(&pool->metadata.lock);
     if (pool->metadata.size_left < aligned_size) {
-        fprintf(stderr,
-                "error: umf_ba_linear_alloc() failed (requested size: %zu > "
-                "space left: %zu)\n",
-                aligned_size, pool->metadata.size_left);
-        util_mutex_unlock(&pool->metadata.lock);
-        return NULL; // out of memory
+        // assert in Debug build (OOM)
+        assert(pool->metadata.size_left >= aligned_size);
+
+        // do NOT fail with OOM in Release build
+        umf_ba_next_linear_pool_t *new_pool =
+            (umf_ba_next_linear_pool_t *)ba_os_alloc(pool->metadata.pool_size);
+        if (!new_pool) {
+            util_mutex_unlock(&pool->metadata.lock);
+            return NULL;
+        }
+
+        void *data_ptr = &new_pool->data;
+        size_t size_left = pool->metadata.pool_size -
+                           offsetof(umf_ba_next_linear_pool_t, data);
+        align_ptr_size(&data_ptr, &size_left, MEMORY_ALIGNMENT);
+
+        pool->metadata.data_ptr = data_ptr;
+        pool->metadata.size_left = size_left;
+
+        // add the new pool to the list of pools
+        new_pool->next_pool = pool->next_pool;
+        pool->next_pool = new_pool;
+#ifndef NDEBUG
+        pool->metadata.n_pools++;
```

Comment:
Will this ever happen? We are failing in Debug on line 111 if there's not enough memory in the pool, right?